### PR TITLE
RPG: Fix regression in smooth lighting animation on moving pit platforms.  Other lighting fixes and improvements.

### PR DESCRIPTION
**Pull request by mrimer**
[Relevant topic](https://forum.caravelgames.com/viewtopic.php?TopicID=44947)

Multiple changes here.  Check out the commit diffs individually for the recommended way to follow the changes.

If this looks good to you, I'll migrate these fixes to 5.1.1.

### CODE REVIEW

**Comment by EvidentlyCube** [comment]
Seems fine, I think I'll have to see it migrated to DROD codebase to better understand all the changes.

**Comment by EvidentlyCube** [comment]
I am reviewing it though! That was just first impression comment!

File: FrontEndLib/BitmapManager.h [hunk 1]
```diff
@@ -68,6 +68,15 @@ struct CustomTiles
 	UINT wTileStartNo;     //starting number of these custom tiles
 };
 
+struct TileMask
+{
+	TileMask(UINT tile, int xOffset, int yOffset)
+		: tile(tile), xOffset(xOffset), yOffset(yOffset)
+	{ }
+	UINT tile;
```
Comment (EvidentlyCube) [inline FrontEndLib/BitmapManager.h:76]: Maybe `tileImage`? Though looking at the rest of the code, TI's are usually referred to as tiles.

File: FrontEndLib/BitmapManager.h [hunk 2]
```diff
@@ -68,6 +68,15 @@ struct CustomTiles
...
+	int xOffset, yOffset; //tile location offset from top-left origin
```
Comment (EvidentlyCube) [inline FrontEndLib/BitmapManager.h:77]: Top-left origin meaning top-left of the room widget? surface? screen? tile?

I am guessing tile, can you add a tiny comment at the top of the struct explaining how it's supposed to be used?

**Review by EvidentlyCube — APPROVED** [review]
Approving but I'd like a few more comments where I mentioned it. I think I get now how it's supposed to work, though I do worry a little bit that it's a very specific solution, and if we ever end up needing something similar for anything else it'll be a lot of work to make it work. But then again I think it's high time to get 5.1.1 out the door, let's avoid refactor creep.

File: FrontEndLib/BitmapManager.h (hunk 1 above)
Comment (mrimer) [reply FrontEndLib/BitmapManager.h:76]: Agree this should be renamed to be more descriptive.
I'm renaming it to TweeningTileMask to indicate it's used for tiles in motion.

File: FrontEndLib/BitmapManager.h (hunk 2 above)
Comment (mrimer) [reply FrontEndLib/BitmapManager.h:77]: Good call out.  I've clarified the comment:
`//relative tile location offset from top-left origin of the area this mask is being applied to`

**Comment by mrimer** [comment]
> I do worry a little bit that it's a very specific solution, and if we ever end up needing something similar for anything else it'll be a lot of work to make it work.

This is a good call out, and I don't disagree.  This was a fair amount of work, and it is a specific solution.  Still, we might find the new routine in CBitmapManager to be something that can be recycled for future complex rendering operations, even if we stop using it for this particular use case.

Fortunately, my solution to fixing the pit darkening is localized to one spot in CRoomWidget, so it would be easy to remove.  If we need to generalize this type of operation (or go with something completely different) in the future, I have no qualms (in principle) about ripping that out and replacing it.

**Comment by mrimer** [comment]
BTW, to check out the light rendering changes, they are easiest to observe if you slow down the key repeat rate and try moving around in a very dark room.  Try also stepping down from a wall or top of a door for z-axis light tweening!  This will also all work when migrated to 5.x.

**Comment by mrimer** [comment]
Ready to review my latest commit.

**Pull request merged**